File: packages/runtime-python/src/eventgen_runtime/adapters/kafka_source.py
from __future__ import annotations

import json
from typing import Any

from eventgen_runtime.types import EventEnvelope, EventMetadata, ReceiveOptions


class _KafkaReceivedMessage:
    def __init__(self, source: "KafkaMessageSource", record: Any, envelope: EventEnvelope) -> None:
        self.envelope = envelope
        self.raw = record
        self.attributes = {
            "topic": record.topic,
            "partition": record.partition,
            "offset": record.offset,
            "headers": record.headers,
        }
        self._source = source
        self._record = record

    async def ack(self) -> None:
        consumer = await self._source.get_consumer()
        await consumer.commit({
            self._record.topic: {
                self._record.partition: self._record.offset + 1,
            }
        })

    async def retry(self, error: Any | None = None) -> None:
        return None

    async def dead_letter(self, error: Any | None = None) -> None:
        await self.retry(error)


class KafkaMessageSource:
    def __init__(
        self,
        topic: str,
        bootstrap_servers: str | list[str],
        group_id: str,
        client_id: str = "eventgen",
    ) -> None:
        self._topic = topic
        self._bootstrap_servers = bootstrap_servers if isinstance(bootstrap_servers, str) else ",".join(bootstrap_servers)
        self._group_id = group_id
        self._client_id = client_id
        self._consumer = None

    async def get_consumer(self):
        if self._consumer is None:
            from aiokafka import AIOKafkaConsumer  # type: ignore[import]

            self._consumer = AIOKafkaConsumer(
                self._topic,
                bootstrap_servers=self._bootstrap_servers,
                group_id=self._group_id,
                client_id=self._client_id,
                enable_auto_commit=False,
            )
            await self._consumer.start()
        return self._consumer

    async def receive(self, options: ReceiveOptions | None = None) -> list[_KafkaReceivedMessage]:
        consumer = await self.get_consumer()
        max_messages = max(1, options.max_messages if options and options.max_messages else 1)
        batch = await consumer.getmany(max_records=max_messages, timeout_ms=1000)
        messages: list[_KafkaReceivedMessage] = []
        for records in batch.values():
            for record in records:
                messages.append(_KafkaReceivedMessage(self, record, _parse_envelope(record.value)))
        return messages

    async def close(self) -> None:
        if self._consumer is not None:
            await self._consumer.stop()
            self._consumer = None


def _parse_envelope(value: bytes | bytearray | str) -> EventEnvelope:
    raw = value.decode("utf-8") if isinstance(value, (bytes, bytearray)) else value
    data = json.loads(raw)
    metadata = data["metadata"]
    return EventEnvelope(
        event_id=data["eventId"],
        version=data["version"],
        payload=data["payload"],
        metadata=EventMetadata(
            created_at=metadata["createdAt"],
            trace_id=metadata["traceId"],
            fifo=metadata.get("fifo"),
        ),
    )

File: packages/runtime-typescript/src/adapters/kafka-source.ts
import { Kafka, type Consumer, type EachMessagePayload, type KafkaMessage } from "kafkajs";
import type { EventEnvelope, MessageSource, ReceiveOptions, ReceivedMessage } from "../types.js";

export type KafkaMessageSourceOptions = {
  topic: string;
  groupId: string;
  brokers: string[];
  clientId?: string;
  kafka?: Kafka;
  fromBeginning?: boolean;
};

type QueuedKafkaMessage = {
  payload: EachMessagePayload;
};

export class KafkaMessageSource implements MessageSource {
  private readonly consumer: Consumer;
  private readonly topic: string;
  private readonly fromBeginning: boolean;
  private readonly queue: QueuedKafkaMessage[] = [];
  private connected = false;
  private closed = false;
  private waiters: Array<() => void> = [];

  constructor(options: KafkaMessageSourceOptions) {
    this.topic = options.topic;
    this.fromBeginning = options.fromBeginning ?? false;
    const kafka = options.kafka ?? new Kafka({
      clientId: options.clientId ?? "eventgen",
      brokers: options.brokers,
    });
    this.consumer = kafka.consumer({ groupId: options.groupId });
  }

  async receive(options?: ReceiveOptions): Promise<ReceivedMessage[]> {
    await this.ensureStarted();
    const maxMessages = Math.max(1, options?.maxMessages ?? 1);

    if (this.queue.length === 0 && !this.closed) {
      await new Promise<void>((resolve) => this.waiters.push(resolve));
    }

    return this.queue.splice(0, maxMessages).map((queued) => this.toReceivedMessage(queued.payload));
  }

  async close(): Promise<void> {
    this.closed = true;
    this.flushWaiters();
    if (this.connected) {
      await this.consumer.disconnect();
      this.connected = false;
    }
  }

  private async ensureStarted(): Promise<void> {
    if (this.connected) return;
    await this.consumer.connect();
    await this.consumer.subscribe({ topic: this.topic, fromBeginning: this.fromBeginning });
    await this.consumer.run({
      autoCommit: false,
      eachMessage: async (payload) => {
        this.queue.push({ payload });
        this.flushWaiters();
      },
    });
    this.connected = true;
  }

  private toReceivedMessage(payload: EachMessagePayload): ReceivedMessage {
    const envelope = parseKafkaEnvelope(payload.message);
    return {
      envelope,
      raw: payload,
      attributes: {
        topic: payload.topic,
        partition: payload.partition,
        offset: payload.message.offset,
        headers: payload.message.headers,
      },
      ack: async () => {
        await this.consumer.commitOffsets([
          {
            topic: payload.topic,
            partition: payload.partition,
            offset: nextOffset(payload.message.offset),
          },
        ]);
      },
      retry: async () => {
        // Keep the offset uncommitted so the consumer group can retry according to Kafka semantics.
      },
    };
  }

  private flushWaiters(): void {
    const waiters = this.waiters.splice(0);
    for (const waiter of waiters) waiter();
  }
}

function parseKafkaEnvelope(message: KafkaMessage): EventEnvelope {
  const value = message.value?.toString("utf8");
  if (!value) throw new Error("Kafka message value is empty");
  const parsed = JSON.parse(value) as EventEnvelope;
  if (!parsed.eventId || !parsed.version || !parsed.metadata) {
    throw new Error("Kafka message value is not an EventEnvelope");
  }
  return parsed;
}

function nextOffset(offset: string): string {
  return (BigInt(offset) + 1n).toString();
}
